core/store/src/trie/mem/arena/mod.rs

```diff
@@ -1,34 +1,83 @@
 mod alloc;
 mod metrics;
+
 use self::alloc::Allocator;
+use super::flexible_data::encoding::BorshFixedSize;
 use borsh::{BorshDeserialize, BorshSerialize};
+use derive_where::derive_where;
 use std::fmt::{Debug, Display, Formatter};
 use std::hash::Hash;
 use std::mem::size_of;
 
-use super::flexible_data::encoding::BorshFixedSize;
-
-/// Arena to store in-memory trie nodes.
+/// Single-threaded `Arena` to store in-memory trie nodes.
 /// Includes a bump allocator that can free nodes.
 ///
 /// To allocate, deallocate, or mutate any allocated memory, a mutable
-/// reference to the `Arena` is needed.
-pub struct Arena {
-    memory: ArenaMemory,
+/// reference to the `STArena` is needed.
+pub struct STArena {
+    memory: STArenaMemory,
     allocator: Allocator,
 }
 
-/// Mmap-ed memory to host the in-memory trie nodes.
+/// An abstraction of an arena that also allows being implemented differently,
+/// specifically in the case of a multi-threaded arena where each thread
+/// behaves like an `STArena` but they share memory chunk allocations.
+pub trait Arena: Sized + 'static {
+    type Memory: ArenaMemory;
+    fn memory(&self) -> &Self::Memory;
+    fn memory_mut(&mut self) -> &mut Self::Memory;
+    /// Allocates a slice of the given size in the arena.
+    fn alloc(&mut self, size: usize) -> ArenaSliceMut<Self::Memory>;
+}
+
+/// The deallocation part of the arena interface; separated from the `Arena`
+/// trait because the multi-threaded implementation does not support deallocation.
+pub trait ArenaWithDealloc: Arena {
```

Comment:
Observation: adding the ability to dealloc arbitrary slices makes this a proper Allocator and could implement the `std::alloc::Allocator` trait if ever needed...

Reply:
It cannot dealloc arbitrary slices and there's no way it ever would, because the arena pointers are two u32 integers rather than a machine pointer.